utils/merge.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package utils
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// Merge will return a new struct of the same type as base and patch, with patch merged into base.
+// Specifically, patch's values will be preferred except when patch's value is `nil`.
+// Restrictions/guarantees:
+//   - base or patch will not be modified
+//   - base and patch can be pointers or values
+//   - base and patch must be the same type
+//   - base and patch must have no unexported types (at the moment)
+//   - if maps or slices are different, the entire map or slice will be replaced (at the moment)
+//
+// Usage: callers need to cast the returned interface back into the original type, eg:
+// func mergeTestStruct(base, patch *testStruct) (*testStruct, error) {
+//	   ret, err := merge(base, patch)
+//	   if err != nil {
+//         return nil, err
```

Comment:
nit: just a comment, but can you indent this for consistency?

Reply:
I thought I had!  (I'm bothered by that kind of thing too...)
Turns out, some of those are tabs, and my editor makes it looked aligned correctly. Will replace with spaces.